src/Npgsql/NpgsqlConnector.cs

```diff
@@ -778,18 +778,47 @@ async Task ConnectAsync(NpgsqlTimeout timeout, CancellationToken cancellationTok
                 Log.Trace($"Attempting to connect to {endpoint}");
                 var protocolType = endpoint.AddressFamily == AddressFamily.InterNetwork ? ProtocolType.Tcp : ProtocolType.IP;
                 var socket = new Socket(endpoint.AddressFamily, SocketType.Stream, protocolType);
+                CancellationTokenSource? combinedCts = null;
+                CancellationTokenSource? timeoutCts = null;
                 try
                 {
+                    // .NET 5.0 added cancellation support to ConnectAsync, which allows us to implement real
+                    // cancellation and timeout. On older TFMs, we fake-cancel the operation, i.e. stop waiting
+                    // and raise the exception, but the actual connection task is left running.
+
+// TODO: NET5_0 is here because of https://github.com/dotnet/sdk/issues/13377, remove for 5.0.0-rc2
+#if (NET461 || NETSTANDARD2_0 || NETSTANDARD2_1 || NETCOREAPP3_1) && !NET5_0
                     await socket.ConnectAsync(endpoint)
                         .WithCancellationAndTimeout(perIpTimeout, cancellationToken);
+#else
+                    var finalCt = cancellationToken;
+
+                    if (perIpTimeout.IsSet)
+                    {
+                        timeoutCts = new CancellationTokenSource(perIpTimeout.TimeLeft.Milliseconds);
+                        finalCt = timeoutCts.Token;
+
+                        if (cancellationToken.CanBeCanceled)
+                        {
+                            combinedCts = CancellationTokenSource.CreateLinkedTokenSource(
+                                cancellationToken, timeoutCts.Token);
+                            finalCt = combinedCts.Token;
+                        }
```

Comment:
```suggestion
                        timeoutCts = new CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
                        timeoutCts.CancelAfter(perIpTimeout.TimeLeft.Milliseconds);
                        finalCt = timeoutCts.Token;
```

Reply:
Good thinking - I'll do the same for the `ReadBuffer` and the `WriteBuffer`.

Reply:
Or not - there, we're reusing the timeout cancellation token source between reads and writes. Unlike here, where it's done only once.

Reply:
What am I missing? CreatedLinkedTokenSource requires at least two tokens in order to be meaningful, no? The code above doesn't compile...

Reply:
It should compile - there is an overload, which takes `params CancellationToken[] tokens`.

Reply:
And for .net 5 they've added another overload, which takes a single `CancellationToken`.

Reply:
Taking another look, it might not compile because of unnecessary `new` here:
```csharp
timeoutCts = new CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
```

Reply:
Ah, I think there was a `new` there before.

I understand now... that's clever... Note that there's a heap-allocated array because of the params, but the previous version has the timeout CTS.

Reply:
Not for new target frameworks.